sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -137,6 +137,10 @@ object Cast {
     case (_: CalendarIntervalType, StringType) => true
     case (NullType, _) => true
 
+    case (CharType(l1), CharType(l2)) => l1 == l2
+    case (CharType(l1), VarcharType(l2)) => l1 <= l2
+    case (VarcharType(l1), VarcharType(l2)) => l1 <= l2
```

Comment:
Could you add tests in `CastSuite`, too? cc: @gengliangwang

Reply:
Can we avoid changing `canUpcast`? We intentionally try to avoid involving char/varchar in cast. Let's create a `canAlterColumnType` somewhere which calls `canUpcast` and deals with char/varchar.

Reply:
OK